### PR TITLE
Sprint 12

### PR DESCRIPTION
**Pull request by Vladlukhanin**
Added:
- Address book functionality;
 - Last activity (https://xmpp.org/extensions/xep-0012.html);
  ```javascript
  QB.chat.getLastUserActivity(userId) - to send query;
  QB.chat.onLastUserActivityListener(userId, seconds) - get last user activity;
```
- Support Safari 11 (experimental);

Updated:
 - API Reference:
   - QB.content, QB.data, QB.user, QB.pushnotifications.
 - possibilty to get user by phone number;

Removed:
 - QB.chat.privacyList.setAsActive();
 - QB.content.taggetForCurrentUser - unnecessary API Call;

Samples:
 - WebRTC sample - Hide record button in if MediaRecorder is unavailable;

How should this be manually tested?
Use Jasmine, start ./spec/SpecRunner.html in browser

Does the documentation need an update?
Yes, generate JSDOC before release

### CODE REVIEW

File: src/modules/AddressBook.js [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+ * @namespace QB.addressbook
+ */
+function AddressBook(service) {
+  this.service = service;
+}
+
+AddressBook.prototype = {
+  /**
+   * The method is used to create, update and delete contacts in address book.<br />
+   * If contacts doesn't exist in address book then it will be created. If contacts exists then it will be updated.
```
Comment (soulfly) [inline src/modules/AddressBook.js]: If contacts -> If contact

File: src/modules/AddressBook.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    return err.code === 404 && errDetails[0] === 'Empty address book';
+  },
+
+  /**
+   * Retrive all contacts from address book.
+   * The method accepts 1 or 2 parameters.
+   * @memberof QB.addressbook
+   * @param {string|function} udidOrCallback - You could pass udid of address book or
+   * callback function if you want to get contacts from global address book.
+   * @param {function} [callback] - Callback function is uses as 2nd parameter if you pass udid as 1st parameters.
```
Comment (soulfly) [inline src/modules/AddressBook.js]: is uses ->  is used

File: src/modules/AddressBook.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        cb(null, res);
+      }
+    });
+  },
+
+  /**
+   * Retrive all user that has phone number from your address book.
+   * The methods accepts 1 or 2 parameters.
+   * @memberof QB.addressbook
+   * @param {boolean|function} udidOrCallback - You can pass isCompact parameter or callback object. If isCompact is passed then only user's id and phone fields will be returned from server. Otherwise - all standard user's parameters will be returned.
```
Comment (soulfly) [inline src/modules/AddressBook.js]: user's parameters -> user's fields

File: src/modules/AddressBook.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+   * @param {function} [callback] - Callback function is uses as 2nd parameter if you pass `isCompact` as 1st parameters.
```
Comment (soulfly) [inline src/modules/AddressBook.js]: is uses -> is used
as 1st parameters. -> as 1st parameter.

File: src/modules/AddressBook.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+          cb(err, null);
+        }
+      } else {
+        cb(null, res);
+      }
+    });
+  },
+
+  /**
+   * Retrive all user that has phone number from your address book.
```
Comment (soulfly) [inline src/modules/AddressBook.js]: Retrive all user that has phone number from your address book. -> Retrieve QuickBlox users that have phone numbers from your address book.

**Comment by soulfly** [comment]
Just a quick question:

Why you decided to have a separate callback for Last Activity instead of the following? :

```
QB.chat.getLastUserActivity(userId, function(seconds, error){

});
```

**Comment by Vladlukhanin** [comment]
"Why you decided to have a separate callback for Last Activity instead of the following?"

To keep one-sided, because the whole chat is working on listeners.

**Pull request merged**